src/main/java/org/testng/internal/BaseTestMethod.java

```diff
@@ -430,38 +430,59 @@ protected void initGroups(Class<? extends ITestOrConfiguration> annotationClass)
     //
     // Init groups depended upon
     //
-    {
-      ITestOrConfiguration annotation = getAnnotationFinder().findAnnotation(getConstructorOrMethod(), annotationClass);
-      ITestOrConfiguration classAnnotation = getAnnotationFinder().findAnnotation(getConstructorOrMethod().getDeclaringClass(), annotationClass);
+    initRestOfGroupDependencies(annotationClass);
+  }
 
-      Map<String, Set<String>> xgd = calculateXmlGroupDependencies(m_xmlTest);
-      List<String> xmlGroupDependencies = Lists.newArrayList();
-      for (String g : getGroups()) {
-        Set<String> gdu = xgd.get(g);
-        if (gdu != null) {
-          xmlGroupDependencies.addAll(gdu);
-        }
+  protected void initBeforeAfterGroups(Class<? extends ITestOrConfiguration> annotationClass, String[] groups) {
+    String[] groupsAtMethodLevel = calculateGroupsTouseConsideringValuesAndGroupValues(annotationClass, groups);
+    //@BeforeGroups and @AfterGroups annotation cannot be used at Class level. So its always null
+    String[] groupsAtClassLevel = null;
+    setGroups(getStringArray(groupsAtMethodLevel, groupsAtClassLevel));
+    initRestOfGroupDependencies(annotationClass);
+  }
+
+  private String[] calculateGroupsTouseConsideringValuesAndGroupValues(Class<? extends ITestOrConfiguration> annotationClass, String[] groups) {
+    ITestOrConfiguration annotation = getAnnotationFinder().findAnnotation(getConstructorOrMethod(), annotationClass);

```

Comment:
Move this line into the if statement. We don't need to execute it if groups is null or empty

Reply:
Fixed, amended commit and pushed the changes once again. Please check